### DISCUSSION TITLE
i have a question in my cuda project

### DISCUSSION THREAD

**Discussion by CalebDu** [clangd/clangd#1891] [Q&A]
i use clangd + cmake build my cuda project in vscode.  i set CMAKE_CUDA_ARCHITECTURES in cmakeList. but clangd server seems can not analyize cuda macro, such as "**#if __CUDA_ARCH__ >= 800**".  All code is dim in "#if __CUDA_ARCH__ >= 800" without code completion.
<img width="882" alt="image" src="https://github.com/clangd/clangd/assets/59528230/12881f86-ca6a-43b3-b028-5bf68f72f4f4">
<img width="853" alt="image" src="https://github.com/clangd/clangd/assets/59528230/6eaa432c-78a5-4265-8b2c-92e6cee7f6ea">

anyone can help me figure it out?

**Comment by HighCommander4** [comment]
Have you looked at https://clangd.llvm.org/installation#project-setup? Are you running `cmake` with `-DCMAKE_EXPORT_COMPILE_COMMANDS=1` as discussed there?

**Reply by CalebDu** [reply]
i got "compile_commands.json" file in build  dir.  clangd server load compile database json successfully and generate ".cache" dir

**Reply by HighCommander4** [reply]
Can you share [clangd logs](https://clangd.llvm.org/troubleshooting#gathering-logs) for further investigation?

**Reply by CalebDu** [reply]
following log is **output** of clangd extension in vscode:

I[12:58:21.286] Indexed c++17 standard library (incomplete due to errors): 14047 symbols, 6713 filtered
I[12:58:22.206] <-- shutdown(1)
I[12:58:22.206] --> reply:shutdown(1) 0 ms
I[12:58:22.237] <-- exit
I[12:58:22.237] LSP finished, exiting with status 0
I[12:58:22.271] clangd version 17.0.3 (https://github.com/llvm/llvm-project 888437e1b60011b8a375dd30928ec925b448da57)
I[12:58:22.271] Features: linux+grpc
I[12:58:22.271] PID: 225401
I[12:58:22.271] Working directory: /home/caleb/cuda_advance
I[12:58:22.271] argv[0]: /root/.vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/17.0.3/clangd_17.0.3/bin/clangd
I[12:58:22.271] argv[1]: --header-insertion=never
I[12:58:22.271] Starting LSP over stdin/stdout
I[12:58:22.271] <-- initialize(0)
I[12:58:22.274] --> reply:initialize(0) 2 ms
I[12:58:22.275] <-- initialized
I[12:58:22.280] <-- textDocument/didOpen
E[12:58:22.281] config error at /home/caleb/cuda_advance/.clangd:1:0: Config should be a dictionary
I[12:58:22.282] --> textDocument/publishDiagnostics
I[12:58:22.284] Loaded compilation database from /home/caleb/cuda_advance/build/compile_commands.json
I[12:58:22.284] ASTWorker building file /home/caleb/cuda_advance/src/example/tensorcore_gemm/gemm.cu version 3 with command 
[/home/caleb/cuda_advance/build]
/usr/local/cuda-12.1/bin/nvcc -forward-unknown-to-host-compiler -I/home/caleb/cuda_advance/3rd_party/cutlass/include -I/home/caleb/cuda_advance/3rd_party/cutlass/tools/util/include -I/home/caleb/cuda_advance/3rd_party/cutlass/examples/common -O3 -DNDEBUG -std=c++17 --generate-code=arch=compute_80,code=[compute_80,sm_80] -x cu -rdc=true -c -o src/example/tensorcore_gemm/CMakeFiles/gemm_tc.dir/gemm.cu.o -resource-dir=/root/.vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/17.0.3/clangd_17.0.3/lib/clang/17 -- /home/caleb/cuda_advance/src/example/tensorcore_gemm/gemm.cu
I[12:58:22.286] --> window/workDoneProgress/create(0)
I[12:58:22.287] Enqueueing 6 commands for indexing
I[12:58:22.287] <-- reply(0)
I[12:58:22.287] --> $/progress
I[12:58:22.287] --> $/progress
I[12:58:22.297] --> textDocument/clangd.fileStatus
I[12:58:22.769] --> $/progress
I[12:58:22.769] --> $/progress
I[12:58:23.536] Built preamble of size 17151468 for file /home/caleb/cuda_advance/src/example/tensorcore_gemm/gemm.cu version 3 in 1.24 seconds
I[12:58:23.538] --> workspace/semanticTokens/refresh(1)
I[12:58:23.538] Indexing c++17 standard library in the context of /home/caleb/cuda_advance/src/example/tensorcore_gemm/gemm.cu
I[12:58:23.539] <-- reply(1)
I[12:58:23.622] --> textDocument/publishDiagnostics
I[12:58:23.623] --> textDocument/inactiveRegions
I[12:58:23.623] --> textDocument/clangd.fileStatus
I[12:58:25.344] Indexed c++17 standard library (incomplete due to errors): 14047 symbols, 6713 filtered

**Reply by HighCommander4** [reply]
I would guess this is likely related to the compile command in `compile_commands.json` being a command for the compiler `nvcc` rather than `clang`. Probably there are some options in that `nvcc` command that `clang` does not recognize and which are relevant to determining the correct value of the `__CUDA_ARCH__` macro.

If you're able to compile your code with `clang`, and arrange for the command in `compile_commands.json` to be a `clang` command rather than an `nvcc` command, that is likely to help.

Otherwise, I'll leave it to folks who are more familiar with CUDA to suggest solutions.

**Comment by jamesnulliu** [comment]
In ".clangd", write this at the bottom:

```
CompileFlags:
  Add: 
    - -Wall
    - -Wextra
  Remove:
    - -ccbin
    - -forward-unknown-to-host-compiler
    - --generate-code*
    - -arch*
    - -rdc=true
    - -Xcudafe
    - --diag_suppress=*
    - --expt-relaxed-constexpr
    - --expt-extended-lambda
    - -Xcompiler*
    - -fmodules*
    - -fmodule-mapper*
    - -fdeps-format*

---

If:
  PathMatch: [.*\.cu, .*\.cuh]
CompileFlags:
  Add: 
    - --cuda-gpu-arch=sm_89
```
